<a href="https://colab.research.google.com/github/LuisMig-code/Fazendo-web-scraping-com-a-biblioteca-BeautifulSoup/blob/main/Fazendo_Web_Scrapping_com_B4S.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Web Scraping usando o BeautifulSoup



<img src="https://scrapingant.com/wp-content/uploads/2020/06/beautifulsoup_logo.png">

**No artigo de hoje pretendo fazer uma espécie de tutorial explicando o que é *Web Scraping* , como fazer um e por que ultilizar**

## O que é?

Antes de começar quaisquer código ou explcação mais complexa vale responder a seguinte pergunta: O que é *Web Scraping* ? Em uma tradução literal , Web Scrapping significa "*Raspagem da web*" e de fato é o que fazemos . Em *Web Scraping* nos podemos pegar um site que contenha informações que nos seja pertinente e "raspar" adiquirindo-os . Normalmente o que é feito são coletas automáticas de informações na Internet , geralmente por meio de programas que simulam a navegação humana na Web.

## Como fazer?

Ants de tudo é importante entender um pouco mais da web , como funcionam os sites e sua comunicação com os servidores

> *Web é uma palavra inglesa que significa teia ou rede. O significado de web ganhou outro sentido com o aparecimento da internet. A web passou a designar a rede que conecta computadores por todo mundo, a World Wide Web (WWW)*.

**[fonte](https://www.significados.com.br/web/)**





O mundo da programção surgiu com uma função , automatizar funções e facilitar a vida das pessoas , com relação a web não seria diferente

A *internet* como organismo não representaria nada se não fossem os sites que a compões, basicamento o que dá sentindo a ela são eles. Mas , **o que são sites e como são feitos**? 



> Site é uma coleção de páginas da web organizadas e localizadas em um servidor na rede. Imagine um site como uma casa onde você reúne seus móveis (as informações dele) em cômodos (as páginas dele).

**[fonte](https://www.hostinger.com.br/tutoriais/o-que-e-site)**






Afinal , **como são feitos**?

Todo site é construido com HTML (*HyperText Markup Language*) , ou seja uma linguagem de marcação constituida de tags (isto será importante parao web scraping) , é lá que podemos escrever o conteúdo do site , os títulos , sub-títulos , paragrafos , imagens , etc. De maneira resumida o HTML é como se fosse o esqueleto de um ser humano , sendo a primera base de sustentação do corpo. 

Além de HTML , sites são construídos usando também CSS3 e JavaScript . o CSS dá os músculos e a aparência do site enquanto o Java Scrpit dá o "cérebro" do site. Eles são importante na construção do site , para não dizer ESSENCIAL , mas por hora , para você que não tenha conhecimento prévio sobre *Web development* , o ideal é que você tenha em mente a ideia de HTML


<img src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fi.pinimg.com%2F736x%2F1d%2F92%2Fb3%2F1d92b3e490b01c7bde7dce6cadfb4133.jpg&f=1&nofb=1">

### Como é esse tal "*HTML*" ?

Tendo em mente o conceito de html vamos à prática.

Se lembra de que eu te falei que HTML é uma linguagem de marcação constituidas por **tags**? Pois bem , vamos saber como elas aparecem:

```
<html>
    <head>
        <title>Isso é um título em HTML</title>
    </head>
    <body>
        <h1>Isso é um H1</h1>
        <p>isso é um parágrafo</p>
        <ul class="lista-1">
            <li>item 1 da lista</li>
            <li>item 2 da lista</li>
            <li>item 3 da lista</li>
        </ul>
        <p>A seguir temos uma DIV.</p>
        <div>
            <div class="classe-da-div">
                <h2>Isso é um H2</h2>
                <ul>
                    <li>item 1 da lista 2</li>
                    <li>item 2 da lista 2</li>
                    <li>item 3 da lista 2</li>
                    <li>item 4 da lista 2</li>
                    <li>item 5 da lista 2</li>
                    <li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li>
                </ul>
            </div>
            <br>
        </div>
    </body>
</html>
```



Assim é um HTML. Estranho né? observe bem temos tags. Onde? cada pequena palavras entre  ``` <> </> ``` . Perce que cada tag tem "começo" e "fim" , assinalados pelo "*/* ". Cada uma tem uma função , não irei explicar todas , [aqui tem um link onde você pode ver por si só](http://www.macoratti.net/Cursos/html5/html5_7.htm) , que inclusive trás um explicação mais aprofundada acerca destes. 

<img src="https://tutorial.techaltum.com/images/element.png">

[*fonte*](https://tutorial.techaltum.com/images/element.png)

Vamos a alguns exemplos , cada ``` H1 , H2 , H3 , H4 ,H5 , H6``` representa uma forma de titulação , ou seja o ```h1``` é a maior titulação (e de maior importâncias) e assim sucessivamente , veja a diferença:

<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>



a tag ```<p></p>``` representa um parágrafo:
<p>Isto é um parágrafo</p>


a tag ``` <ul></ul>``` representa uma lista não ordenada , dentro dela estão itens enumerados com a tag ```<il></il>``` 

<p>Lista:</p>
<ul>
  <li>item 1</il>
</ul>

Por último e talvez uma das mais importântes , temos a tag ```<div></div>``` que cria uma divisão . Esta tag é meramente semântica , não trás alterações no texto , mas serve para organizar o código , por exemplo , você pode criar uma div específica para ter uma lista X e dar a ela uma classe (*class*).

O que são *classes*?

<p> Quando se quer dar "estilo" a algum item usando CSS é comum delimitar cada item com uma *class* , por exemplo , caso tenhamos uma *div* apenas para uma determinada lista , e queremos que esta tenha a cor azul e o fundo vermelhor , delimitamos a div com uma classe e no CSS fazemos uma edição visual da seguinte forma:</p>


```
<style>
	.minha-classe {
		color: red;
    background-color = blue ;
	}
</style>
```

perceba bem que colocamos um "." no inicio , ele é necessário para que o CSS saiba que se trata de uma classe


### Nosso Primeiro *Scraping*:

A priori , vamos usar o código html que mostramos acima para começar

In [1]:
html = """
<html>
    <head>
        <title>Isso é um título em HTML</title>
    </head>
    <body>
        <h1>Isso é um H1</h1>
        <p>isso é um parágrafo</p>
        <ul class="lista-1">
            <li>item 1 da lista</li>
            <li>item 2 da lista</li>
            <li>item 3 da lista</li>
        </ul>
        <p>A seguir temos uma DIV.</p>
        <div>
            <div class="classe-da-div">
                <h2>Isso é um H2</h2>
                <ul>
                    <li>item 1 da lista 2</li>
                    <li>item 2 da lista 2</li>
                    <li>item 3 da lista 2</li>
                    <li>item 4 da lista 2</li>
                    <li>item 5 da lista 2</li>
                    <li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li>
                </ul>
            </div>
            <br>
        </div>
    </body>
</html>
"""

Aqui já temos um código HTML completo , já lido por uma variável , mas normalmente você usará um link de um site não um código HTML direto em uma variável , posteriormente ensinarei a ler uma página usando apenas um link

Vamos primeiramente importar tudo o que precisaremos para nossas análises e depois criar nosso Objeto BeatifulSoup:

In [2]:
# Importando pacotes necessários:

from bs4 import BeautifulSoup

antes de tudo vamos ver como está nossa variável html?

In [3]:
html

'\n<html>\n    <head>\n        <title>Isso é um título em HTML</title>\n    </head>\n    <body>\n        <h1>Isso é um H1</h1>\n        <p>isso é um parágrafo</p>\n        <ul class="lista-1">\n            <li>item 1 da lista</li>\n            <li>item 2 da lista</li>\n            <li>item 3 da lista</li>\n        </ul>\n        <p>A seguir temos uma DIV.</p>\n        <div>\n            <div class="classe-da-div">\n                <h2>Isso é um H2</h2>\n                <ul>\n                    <li>item 1 da lista 2</li>\n                    <li>item 2 da lista 2</li>\n                    <li>item 3 da lista 2</li>\n                    <li>item 4 da lista 2</li>\n                    <li>item 5 da lista 2</li>\n                    <li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li>\n                </ul>\n            </div>\n            <br>\n        </div>\n    </body>\n</html>\n'

observe que ela está meio "bagunçada" , existem espaços em excesso e onde temos quebra de linha foi colocado um "\n" . A fim de facilitar nossa análise seria interessante "consertar" isso para evitar quaisquer problemas futuros , para isso vamos criar uma função que faça isso

In [4]:
def trata_HTML(input):
    return " ".join(input.split()).replace('> <','><')

In [5]:
html = trata_HTML(html)

Será que deu certo? vamos ver: 

In [6]:
html

'<html><head><title>Isso é um título em HTML</title></head><body><h1>Isso é um H1</h1><p>isso é um parágrafo</p><ul class="lista-1"><li>item 1 da lista</li><li>item 2 da lista</li><li>item 3 da lista</li></ul><p>A seguir temos uma DIV.</p><div><div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div><br></div></body></html>'

Aparentemente resolveu.

Vamos agora criar nosso objeto *BeautifulSoup* e dar a ele esse HTML

In [7]:
soup = BeautifulSoup(html, 'html.parser')

O que acha de vermos como está nosso objeto *soup* ?

In [8]:
soup

<html><head><title>Isso é um título em HTML</title></head><body><h1>Isso é um H1</h1><p>isso é um parágrafo</p><ul class="lista-1"><li>item 1 da lista</li><li>item 2 da lista</li><li>item 3 da lista</li></ul><p>A seguir temos uma DIV.</p><div><div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div><br/></div></body></html>

aqui temos nossa página html toda contida em uma linha . Vamos começar a de fato colocar a mão na massa e "raspar" essas página ?

#### Find e *FindAll* :

O primeiro método que vamos aprender vai ser o **Find** e o **FindAll** , que como o próprio nome sugere , procura ou procura tudo oque pedirmos a ele. Mas como?

```
find ou findAll (tag_html , class_="sua classe")
```

Esse método pede pelo menos um desse para começar a localizar um item. Vamos testar?



In [9]:
soup.find("div")

<div><div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div><br/></div>

In [10]:
soup.findAll("div")

[<div><div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div><br/></div>,
 <div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div>]

Perceba que o método *find* me retorna apenas o primeiro item que ele achar com os parâmetros que passamos , neste caso ele retornou a **primeira** "div" que ele achou , já o findAll retorna uma **lista com todos os itens procurados**

podemos ser ainda mais específicos

In [11]:
soup.find("div" , class_="classe-da-div")

<div class="classe-da-div"><h2>Isso é um H2</h2><ul><li>item 1 da lista 2</li><li>item 2 da lista 2</li><li>item 3 da lista 2</li><li>item 4 da lista 2</li><li>item 5 da lista 2</li><li class="vencedor">item 6 da lista 2<strong> [NEGRITO]</strong></li></ul></div>

O que estamos "dizendo" para nosso objeto é : *Busque pra mim a primeira tag div que você achar com a classe 'classe-da-div'* 

Digamos que queremos alé, vamos buscar um dos itens da lista , o 2° item

In [12]:
soup.find("div" , class_="classe-da-div").findAll("li")[1]

<li>item 2 da lista 2</li>

Interessante , agora se você perceber bem , o que estamos pegando é a tag html completa como fariamos para pegar o texto dentro dela?

#### get_text()

Este atributo é bem sugestivo e auto-explicativo , o que ele faz é pegar o texto da tag selecionada

In [13]:
soup.find("div" , class_="classe-da-div").findAll("li")[1].get_text()

'item 2 da lista 2'

#### attrs

É comum que tags tragam atributos consigo , ou seja , mais informações, neste caso temos apenas uma classe em nossa div , mas podemos ter outros atributos lá , atributos quaisquer mas que tragam informações pertinentes , mas como retê-los?

In [14]:
soup.find("div" , class_="classe-da-div").attrs

{'class': ['classe-da-div']}

Com o método *attrs* podemos pegar todos os atributos da tag e mostrá-los em um dicionário 

In [15]:
soup.find("div" , class_="classe-da-div").attrs["class"]

['classe-da-div']

### Mão na massa:

É evidente que não pretendo me limitar a um exmplo simples , vamos agora fazer um *Scraping* real de dados da internet usando tudo o que já aprendemos com alguns adicionais

**Os dados que iremos "raspar" serão do site [AutoLine](https://www.autoline.com.br/) e tudo aqui usado é para fins educacionais e não comerciais , não há quaisquer patrocínio neste artigo.**

#### Abrindo o site

Diferentemente do que fizemos anteriormente , é extremamente não-eficiente se copiássemos o html da página que quisermos fazer a "raspagem". Para isso , temos de encontrar um forma de "abrir" no site e ler o html dele. o BeautifulSoup dispõe de uma ferramenta como essa

In [16]:
from urllib.request import urlopen

In [17]:
url = "https://busca.autoline.com.br/comprar/carros/novos-seminovos-usados/todos-os-estados/todas-as-cidades/todas-as-marcas/todos-os-modelos/todas-as-versoes/todos-os-anos/todas-as-cores/todos-os-precos/?page=1"

perceba que no final da nossa URL temos um atributo "?page=1" , é fundamental notar isso , pois é com ele que faremos as iterações para o Web Scraping das várias páginas.


Vamos ver acessar a página agora

In [18]:
resposta = urlopen(url)

acessado a página , vamos agora ler o conteúdo dela

In [19]:
html = resposta.read()

Para o Brasil , o padrão de codificação binária de comprimento variável utilizado é o UTF-8 , por isso vamos converter nosso html para esse padrão

In [20]:
html = html.decode('utf-8')

Como já fizemos acima , vamos tratar o nosso html e tirar os espaçamentos e quebras de linha

In [21]:
html = trata_HTML(html)

#### Iniciando a busca por dados:

Vamos criar um objeto B4S para começar a "navegar" no site

In [22]:
soup = BeautifulSoup(html , 'html.parser')

A fim de evitar mais problemas , já fui no site e avaliei o que estava buscando.

O que buscamos é:


*   div
*   classe = nm-product-item



In [23]:
soup.find(class_="nm-product-item")

<li class="nm-product-item " data-brand="CHEVROLET" data-category="Hatch" data-city="São Paulo" data-name="CHEVROLET JOY" data-pid="13983978" data-price="33960.00" data-state="SP" data-url="//www.autoline.com.br/carro/chevrolet/joy-10-8v-flex-4p-manual/2020/sao-paulo-sp/13983978" id="nm-product-13983978" itemscope="" itemtype="https://schema.org/Product"><!-- FAVORITE --><a class="nm-like-icon no-watch-click" data-pid="13983978" href="#"></a><!-- product image --><div class="nm-product-img-container"><!-- THUMBNAIL --><a class="nm-product-img-link no-watch-click" href="//www.autoline.com.br/carro/chevrolet/joy-10-8v-flex-4p-manual/2020/sao-paulo-sp/13983978" itemprop="url" title="CHEVROLET JOY"><img alt="//www.autoline.com.br/carro/chevrolet/joy-10-8v-flex-4p-manual/2020/sao-paulo-sp/13983978" class="nm-product-img" itemprop="image" onerror="this.setAttribute('src', 'images/sem-foto.png')" src="https://w.cdnautoline.com.br/images/CHEVROLET-JOY-13983978-1.jpg" title="CHEVROLET JOY"/></a

Entretanto , não buscamos apenas a div em si , queremos os atributos que ela carrega consigo , lá estão dados necessários para nossas análise

In [24]:
itens = soup.find(class_="nm-product-item").attrs
itens

{'class': ['nm-product-item', ''],
 'data-brand': 'CHEVROLET',
 'data-category': 'Hatch',
 'data-city': 'São Paulo',
 'data-name': 'CHEVROLET JOY',
 'data-pid': '13983978',
 'data-price': '33960.00',
 'data-state': 'SP',
 'data-url': '//www.autoline.com.br/carro/chevrolet/joy-10-8v-flex-4p-manual/2020/sao-paulo-sp/13983978',
 'id': 'nm-product-13983978',
 'itemscope': '',
 'itemtype': 'https://schema.org/Product'}

Com isso conseguimos um dicionário de dados que trás consigo informações relevantes sobre o veículo , como a marca , categoria , nome , estado da federação e etc. Podemos acessar qualquer item da seguinte forma:



```
dicionário["chave_do_dicionario"]
```

Portanto , mão a obra.


Para cada item que quero irei guardar em uma variável


In [25]:
marca = itens['data-brand']
marca

'CHEVROLET'

In [26]:
cidade = itens["data-city"]
cidade

'São Paulo'

In [27]:
nome = itens['data-name']
nome

'CHEVROLET JOY'

In [28]:
categoria = itens["data-category"]
categoria

'Hatch'

In [29]:
estado = itens['data-state']
estado

'SP'

In [30]:
link = itens['data-url'].replace("//","")
link

'www.autoline.com.br/carro/chevrolet/joy-10-8v-flex-4p-manual/2020/sao-paulo-sp/13983978'

In [31]:
preco = itens['data-price']
preco = float(preco)
preco

33960.0

Caso tenha notado , a forma de acessar o preço é diferente , isso porque quando pegamos qualquer dado no *BeautifulSoup* ele nos retorna no formato de string (texto) , por isso devemos transformá-lo para número inteiro ou ponto flutuante.

Entretanto , não busco apenas estes dados , seria interessante também buscarmos a kilometragem rodada do carro , o ano de fabricação e o ano do modelo.

Previamente procurei os atributos que buscávamos , estes são eles:



*   div
*   classe = "nm-features nm-year" para o ano de fabricação e modelo
*   classe = "nm-features nm-km " para a kilometragem rodada



In [32]:
ano_de_fabricacao = soup.find("div",class_="nm-features nm-year").get_text()
ano_de_fabricacao = int(ano_de_fabricacao.split("/")[0].replace(" ",""))
ano_de_fabricacao

2019

In [33]:
ano_do_modelo = soup.find("div",class_="nm-features nm-year").get_text()
ano_do_modelo = int(ano_do_modelo.split("/")[1].replace(" ",""))
ano_do_modelo

2020

In [34]:
km_rodados = soup.find("div" , class_="nm-features nm-km ").get_text()
km_rodados = int(km_rodados.replace(" ",""))
km_rodados

0

Assim como citado já acima , quando queremos dados em forma de números devemos  realizar transformações na variável

O que acabamos de fazer foi recuperar os dados que temos em um dos cards (o primeiro) da primeira página . O que fazemos para termos todos os dados da primeira página? simples: feremos um rotina *Web Scraping*

#### Rotina Web Scraping para a primeira página:



> Rotina é um sequência de procedimentos, dos costumes habituais


Ou seja , quando falamos que vamos fazer uma "rotina" *Web Scraping* , falamos que faremos repetidas ações/procedimentos de raspagem de dados.

Em resumo , estamos falando de fazer iterações na página , e jogar os dados recolhidos em um *Dataframe* (um tipo de estrutura de dados da biblioteca Pandas)


É fundamental entendermos que para podermos jogar os dados em um Dataframe temos de primeiramente armazenar o que temos em um dicionário python e depois aglutinar todos os dicionários em uma lista.


Para que fique compreensível faremos assim: Em cada item da página teremos um dicionário guardando os dados de **1 único item** , após isso , **juntaremos todos os dicionários em uma lista**

In [42]:
# url da página a ser "raspada"
url = "https://busca.autoline.com.br/comprar/carros/novos-seminovos-usados/todos-os-estados/todas-as-cidades/todas-as-marcas/todos-os-modelos/todas-as-versoes/todos-os-anos/todas-as-cores/todos-os-precos/?page=1"

# abrindo a url
response = urlopen(url)

# lendo o arquivo html da url
html = response.read()

# decodificando para o padrão UTF-8 (Mundial) evitando problemas 
html = html.decode('utf-8')

# Tratando o html (removendo os espaçoes entre as tags)
html = trata_HTML(html)

# criando um objeto B4S com a html tratada
soup = BeautifulSoup(html , 'html.parser')

# Listas a serem preenchidas
dados_gerais = []
dados_fabricacao = []
dados_km_rodados = []

# itens iteráveis da página
itens = soup.findAll(class_="nm-product-item")
item_dados_de_fabricacao = soup.findAll("div",class_="nm-features-container nm-automaker-list-view")
item_dados_de_kilometragem = soup.findAll("div",class_="nm-features-container nm-automaker-list-view")


#iteração dos itens gerais
for item in itens:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # marca do carro
    card['marca'] = item.attrs['data-brand']
    # cidade onde se encontra
    card['cidade'] = item.attrs["data-city"]
    # nome do carro 
    card['nome'] = item.attrs['data-name']
    # tipo do carro(categoria)
    card["categoria"] = item.attrs["data-category"]
    # preço do carro
    preco = item.attrs['data-price']
    card['preco'] = float(preco)
    # estado da federação onde o carro se encontra
    card['estado'] = item.attrs['data-state']
    # link para acessar a página web do carro
    card['link'] = item.attrs['data-url'].replace("//","")

    # jogando os dados do dicionário em uma única lista
    dados_gerais.append(card)

# iteração do modelo e o ano de fabricação
for item in item_dados_de_fabricacao:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # ano de fabricação do carro
    ano_de_fabricacao = item.find(class_="nm-features nm-year").get_text()
    ano_de_fabricacao = int(ano_de_fabricacao.split("/")[0].replace(" ",""))
    card["ano_de_fabricacao"] = ano_de_fabricacao
    
    # ano do modelo do carro
    ano_do_modelo = item.find(class_="nm-features nm-year").get_text()
    ano_do_modelo = int(ano_do_modelo.split("/")[1].replace(" ",""))
    card["ano_do_modelo"] = ano_do_modelo

    dados_fabricacao.append(card)

# iteração da kilometragem dos carros
for item in item_dados_de_kilometragem:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # kilometragem do carro
    km_rodados = item.find(class_="nm-features nm-km ").get_text()
    km_rodados = int(km_rodados.replace(".",""))
    card["km_rodados"] = km_rodados

    dados_km_rodados.append(card)  

Feito isso , temos um lista com os dados gerais , outra guardando a kilometragem e outra guardando o ano de fabricação e o ano do modelo. Com isso , vamos importar a biblioteca do *Pandas* e criar um *DataFrame* com estes dados.

In [43]:
# imports necessários

import pandas as pd

In [44]:
# unindo todos os dados em um único DataFrame 
dados = pd.concat([pd.DataFrame(dados_gerais) , pd.DataFrame(dados_fabricacao) , pd.DataFrame(dados_km_rodados)] , axis=1)

In [45]:
# visualizando os dados
dados

marca          cidade  ... ano_do_modelo km_rodados
0       CHEVROLET       São Paulo  ...          2020          0
1         HYUNDAI       São Paulo  ...          2020          0
2         HYUNDAI       São Paulo  ...          2021          0
3           HONDA       São Paulo  ...          2014      78348
4            FIAT       São Paulo  ...          2012      15990
5       CHEVROLET       São Paulo  ...          2015      75000
6       CHEVROLET       São Paulo  ...          2020          0
7      VOLKSWAGEN       São Paulo  ...          2018      75281
8       CHEVROLET  Rio de Janeiro  ...          2017      72000
9       CHEVROLET       São Paulo  ...          2021          0
10      CHEVROLET       São Paulo  ...          2020          0
11          HONDA       São Paulo  ...          2019       6070
12           FIAT       São Paulo  ...          2018      27000
13      CHEVROLET       São Paulo  ...          2021          0
14        HYUNDAI       São Paulo  ...          2014     109900
15      CHEVROLET       São Paulo  ...          2015      87408
16        RENAULT       São Paulo  ...          2010     156453
17           JEEP       São Paulo  ...          2017      47253
18        HYUNDAI       São Paulo  ...          2016      40009
19           FORD       São Paulo  ...          2013      73000
20        HYUNDAI       São Paulo  ...          2011     161229
21  MERCEDES-BENZ       São Paulo  ...          2020       4100
22        CITROEN       São Paulo  ...          2017      28000
23        RENAULT       São Paulo  ...          2021          0

[24 rows x 10 columns]

Como podemos ver , temos um *Dataframe* com 24 entradas que incluem tudo aquilo que queremos. 

Entretanto , observe que apenas 24 dados não é o suficiente , principalmente quando queremos fazer algum modelo de Machine Learning , para isso será necessários "raspar" os dados de outra páginas , vamos criar outra rotina *Web Scraping*

#### Rotina Web Scraping para outras páginas

A lógica a ser seguida aqui será semelhante a da primeira página , mas dessa vez teremos que fazer 2 iterações , uma dentro da outra , pulando páginas a partir do link , pois no fim de cada link temos um atributo de página que podemos mudar de acordo com a página que queremos:

"*?page=[número da página que queremos]*"

Mas antes de tudo , o que temos de ter em mente é: Quanto mais páginas quisermos , mais processamento será gasto , ou seja , podemo pedir 2 mil páginas mas nos custará mais tempo . O ideal é que você estabeleça um número de páginas a serem raspadas que já contenha dados sufientes para análises , o fato de você ter 4 mil dados ou 4 milhões vai depender da sua necessidade.

Neste primeiro exemplo me limitarei a raspar apenas 100 páginas , o que no fim vai me dar um total de 2400 entradas

In [46]:
# Listas a serem preenchidas
dados_gerais = []
dados_fabricacao = []
dados_km_rodados = []

# itens iteráveis da página
itens = soup.findAll(class_="nm-product-item")
item_dados_de_fabricacao = soup.findAll("div",class_="nm-features-container nm-automaker-list-view")
item_dados_de_kilometragem = soup.findAll("div",class_="nm-features-container nm-automaker-list-view")

# Número de páginas que queremos
paginas = 100

for pagina in range(0,paginas):
  # url da página a ser "raspada"
  url = "https://busca.autoline.com.br/comprar/carros/novos-seminovos-usados/todos-os-estados/todas-as-cidades/todas-as-marcas/todos-os-modelos/todas-as-versoes/todos-os-anos/todas-as-cores/todos-os-precos/?page={}".format(pagina)

  # abrindo a url
  response = urlopen(url)

  # lendo o arquivo html da url
  html = response.read()

  # decodificando para o padrão UTF-8 (Mundial) evitando problemas 
  html = html.decode('utf-8')

  # Tratando o html (removendo os espaçoes entre as tags)
  html = trata_HTML(html)

  # criando um objeto B4S com a html tratada
  soup = BeautifulSoup(html , 'html.parser')

  #iteração dos itens gerais
  for item in itens:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # marca do carro
    card['marca'] = item.attrs['data-brand']
    # cidade onde se encontra
    card['cidade'] = item.attrs["data-city"]
    # nome do carro 
    card['nome'] = item.attrs['data-name']
    # tipo do carro(categoria)
    card["categoria"] = item.attrs["data-category"]
    # preço do carro
    preco = item.attrs['data-price']
    card['preco'] = float(preco)
    # estado da federação onde o carro se encontra
    card['estado'] = item.attrs['data-state']
    # link para acessar a página web do carro
    card['link'] = item.attrs['data-url'].replace("//","")

    # jogando os dados do dicionário em uma única lista
    dados_gerais.append(card)

  # iteração do modelo e o ano de fabricação
  for item in item_dados_de_fabricacao:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # ano de fabricação do carro
    ano_de_fabricacao = item.find(class_="nm-features nm-year").get_text()
    ano_de_fabricacao = int(ano_de_fabricacao.split("/")[0].replace(" ",""))
    card["ano_de_fabricacao"] = ano_de_fabricacao
    
    # ano do modelo do carro
    ano_do_modelo = item.find(class_="nm-features nm-year").get_text()
    ano_do_modelo = int(ano_do_modelo.split("/")[1].replace(" ",""))
    card["ano_do_modelo"] = ano_do_modelo

    dados_fabricacao.append(card)

  # iteração da kilometragem dos carros
  for item in item_dados_de_kilometragem:
    #criando um dicionário vazio para adicionar gradualmente as infos do carro
    card = {}

    # kilometragem do carro
    km_rodados = item.find(class_="nm-features nm-km ").get_text()
    km_rodados = int(km_rodados.replace(".",""))
    card["km_rodados"] = km_rodados

    dados_km_rodados.append(card)  

In [47]:
# unindo todos os dados em um único DataFrame 
dados = pd.concat([pd.DataFrame(dados_gerais) , pd.DataFrame(dados_fabricacao) , pd.DataFrame(dados_km_rodados)] , axis=1)

In [49]:
# visualisando os dados
dados.head()

marca     cidade  ... ano_do_modelo km_rodados
0  CHEVROLET  São Paulo  ...          2020          0
1    HYUNDAI  São Paulo  ...          2020          0
2    HYUNDAI  São Paulo  ...          2021          0
3      HONDA  São Paulo  ...          2014      78348
4       FIAT  São Paulo  ...          2012      15990

[5 rows x 10 columns]

In [50]:
# quantidade de dados:
print("temos {} colunas e {} linhas".format(dados.shape[1],dados.shape[0]))

temos 10 colunas e 2400 linhas


#### Exportando nossos dados:

Após termos criado um *DataFrame* , é importante que saibamos exportar estes dados e fazer o download do mesmo para nossa máquina , a fim de posteriormente usar eles.

obs: Aqui estou usando o GoogleColab , o que significa que após exportar tenho de baixá-lo , mas caso você esteja usando o Jupyter , ele automaticamente aloca o dado na sua máquina

In [56]:
# exportando :
dados.to_csv('cars.csv',index=False,encoding='utf-8-sig')

In [57]:
# Visualisando  os dados que exportamos
pd.read_csv("cars.csv")

marca     cidade  ... ano_do_modelo km_rodados
0         CHEVROLET  São Paulo  ...          2020          0
1           HYUNDAI  São Paulo  ...          2020          0
2           HYUNDAI  São Paulo  ...          2021          0
3             HONDA  São Paulo  ...          2014      78348
4              FIAT  São Paulo  ...          2012      15990
...             ...        ...  ...           ...        ...
2395           FORD  São Paulo  ...          2013      73000
2396        HYUNDAI  São Paulo  ...          2011     161229
2397  MERCEDES-BENZ  São Paulo  ...          2020       4100
2398        CITROEN  São Paulo  ...          2017      28000
2399        RENAULT  São Paulo  ...          2021          0

[2400 rows x 10 columns]

Aparentemente está tudo OK , mas por garantia vamos ver se temos algum dados nulo e os tipos de dados neste *DataFrame*

In [58]:
data = pd.read_csv("cars.csv")
data.isnull().sum()

marca                0
cidade               0
nome                 0
categoria            0
preco                0
estado               0
link                 0
ano_de_fabricacao    0
ano_do_modelo        0
km_rodados           0
dtype: int64

Nenhum dado nulo

In [59]:
data.dtypes

marca                 object
cidade                object
nome                  object
categoria             object
preco                float64
estado                object
link                  object
ano_de_fabricacao      int64
ano_do_modelo          int64
km_rodados             int64
dtype: object

Os tipos dos nossos dados também estão corretos

#### **Bônus** (Scraping de TODAS as páginas)

Digamos que você necessite de fato de todas as páginas , mas quer , de forma automatzada , recolher todas elas , ensinarei aqui como fazer isso.

In [63]:
# obtendo a url da página 1
path = "https://busca.autoline.com.br/comprar/carros/novos-seminovos-usados/todos-os-estados/todas-as-cidades/todas-as-marcas/todos-os-modelos/todas-as-versoes/todos-os-anos/todas-as-cores/todos-os-precos/"

# abrindo a url
response = urlopen(url)

#decodificando o html para UTF-8
html = response.read().decode('utf-8')

# Criando o objeto BeautifulSoup
soup = BeautifulSoup(html , 'html.parser')

#descobrindo o número de itens totais
n_itens = soup.find("h3", class_="sub-title").find(class_="neemu-total-products-container").find("strong").get_text().replace(".","").split()[0]
n_itens = int(n_itens)

print(n_itens)

165751


Observe que o que descobrimos não foi o número de páginas , mas sim o total de carros cadastrados , que o próprio site disponibiliza . Para descobrir as páginas , devemos dividir este valor por 24 , ou seja o número de itens por páginas

In [64]:
# dividindo o número de carros pela quantidade de itens por página
páginas = int(n_itens / 24)
print(páginas)

6906


Assim , descobrimos que temos na verdade 6906 páginas , caso você ache ser necessário ter todos os dados dos carros , você pode usar este valor nas iterações

## Conclusões:

*Web Scraping* é uma técninca fundamental para cientista de dados e que cada vez mais vem sendo reconhecida como tal. Muitas empresas quando começam necessitam deste profissional e por vezes a mesma não tem dados o suficiente , portanto , cabe ao cientista de dados coletar eles no processo de "**Mineração de Dados**" e realizar a tarefa proposta . A ideia de coletar dados por servir para inumeras coisas , Marketing , análises de comportamento , migrar dados para outro site ou mesmo para um profissional que está começando e busca dados mais "originais". Pense mais nesta ideia , busque mais conhecimento sobre esta skill e com certeza será um grande profissional.